### PR TITLE
CybersourceREST - Refund | Credit

### PR DESCRIPTION
**Pull request by sinourain**
Description
-------------------------
This integration support the following payment operations:
- Refund
- Credit

Unit test
-------------------------
Finished in 37.39707 seconds.
5452 tests, 77126 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications 100% passed

Remote test
-------------------------
145.79 tests/s, 2062.35 assertions/s

Rubocop
-------------------------
760 files inspected, no offenses detected

GWI-471

Important note: 
Previous review here: https://github.com/activemerchant/active_merchant/pull/4696

### CODE REVIEW

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb [hunk 1]
```diff
@@ -129,8 +156,17 @@ def add_address(post, payment_method, address, options, address_type)
...
+
+        merchant = post[:merchantInformation][:merchantDescriptor] = {}
+        merchant[:name] = options[:merchant_descriptor_name] if options[:merchant_descriptor_name]
+        merchant[:address1] = options[:merchant_descriptor_address1] if options[:merchant_descriptor_address1]
+        merchant[:locality] = options[:merchant_descriptor_locality] if options[:merchant_descriptor_locality]
+      end
+
       def url(action)
-        "#{(test? ? test_url : live_url)}#{action}"
+        "#{(test? ? test_url : live_url)}/pts/v2/#{action}"
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:169]: This seems like a big change - was it something that we didn't catch for the initial implementation, or is this something that is only required for refunds/credits?

Comment (curiousepic) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:169]: Do we expect to need to hit a different endpoiont for other types of actions later? If not, I think we can just add it to the base `self.live/test_url`s

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb [hunk 2]
```diff
@@ -44,7 +44,17 @@ def authorize(money, payment, options = {}, capture = false)
...
+      end
+
+      def refund(money, authorization, options = {})
+        post = build_refund_request(money, options)
+        commit("payments/#{authorization}/refunds", post)
+      end
+
+      def credit(money, payment, options = {})
+        post = build_credit_request(money, payment, options)
+        commit('credits', post)
```
Comment (curiousepic) [inline lib/active_merchant/billing/gateways/cyber_source_rest.rb:57]: Should this be `credits/`?

File: test/remote/gateways/remote_cyber_source_rest_test.rb [hunk 3]
```diff
@@ -70,6 +70,88 @@ def test_successful_purchase
...
+    response = @gateway.refund(@amount / 2, purchase.authorization, @options)
+
+    assert_success response
+    assert response.test?
+    assert_equal 'PENDING', response.message
+    assert response.params['id'].present?
+    assert response.params['_links']['void'].present?
+  end
+
+  def test_successful_more_than_the_payment_amount_refund
```
Comment (curiousepic) [inline test/remote/gateways/remote_cyber_source_rest_test.rb:105]: Shouldn't we expect this to fail?

File: test/remote/gateways/remote_cyber_source_rest_test.rb [hunk 4]
```diff
@@ -70,6 +70,88 @@ def test_successful_purchase
...
+    purchase = @gateway.purchase(@amount, @card_without_funds, @options)
+    response = @gateway.refund(@amount, purchase.authorization, @options)
+
+    assert_failure response
+    assert response.test?
+    assert_match %r{Declined - One or more fields in the request contains invalid data}, response.params['message']
+    assert_equal 'INVALID_DATA', response.params['reason']
+  end
+
+  def test_successful_less_than_the_payment_amount_refund
```
Comment (curiousepic) [inline test/remote/gateways/remote_cyber_source_rest_test.rb:94]:
```suggestion
  def test_successful_partial_refund
```

File: test/remote/gateways/remote_cyber_source_rest_test.rb (hunk 3 above)
Comment (curiousepic) [reply test/remote/gateways/remote_cyber_source_rest_test.rb:105]: I guess if their sandbox simulates the delay in confirming, we wouldn't, and just get a pending result.

File: test/remote/gateways/remote_cyber_source_rest_test.rb [hunk 5]
```diff
@@ -70,6 +70,88 @@ def test_successful_purchase
...
+    response = @gateway.refund(@amount * 2, purchase.authorization, @options)
+
+    assert_success response
+    assert response.test?
+    assert_equal 'PENDING', response.message
+    assert response.params['id'].present?
+    assert response.params['_links']['void'].present?
+  end
+
+  def test_successful_double_refund_transactions
```
Comment (curiousepic) [inline test/remote/gateways/remote_cyber_source_rest_test.rb:116]:
```suggestion
  def test_successful_repeat_refund_transaction
```

Comment (curiousepic) [reply test/remote/gateways/remote_cyber_source_rest_test.rb:116]: Also something we should expect to fail, but I guess doesn't if sandbox simulates pending state...

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb (hunk 1 above)
Comment (sinourain) [reply lib/active_merchant/billing/gateways/cyber_source_rest.rb:169]: Yes, reviewing the changes from @scastilloendava looks like the paths for [store/unstore](https://github.com/activemerchant/active_merchant/pull/4709) have another subpath

File: lib/active_merchant/billing/gateways/cyber_source_rest.rb (hunk 2 above)
Comment (sinourain) [reply lib/active_merchant/billing/gateways/cyber_source_rest.rb:57]: No. Thanks, I removed  `/` in the `payments` method

**Review by curiousepic — APPROVED** [review]
👍  Nice work, merging.

**Comment by curiousepic** [comment]
@sinourain oops, could you squash the commits before I merge?

File: test/remote/gateways/remote_cyber_source_rest_test.rb (hunk 3 above)
Comment (sinourain) [reply test/remote/gateways/remote_cyber_source_rest_test.rb:105]: At this moment I have no way to prove that a refund of greater value than the purchase is complete (even if it should fail), I get to the pending response therefore I will remove this test

**Comment by sinourain** [comment]
> @sinourain oops, could you squash the commits before I merge?

Yes, I will, no worries

**Pull request closed without merging**